docker/Dockerfile

```diff
@@ -1,34 +1,127 @@
-ARG BASE_IMAGE=python:3.8-slim-bullseye
+##
+#  Generic dockerfile for dbt image building.  
+#  See README for operational details
+##
 
-FROM $BASE_IMAGE
-ARG BASE_REQUIREMENTS_SRC_PATH
-ARG WHEEL_REQUIREMENTS_SRC_PATH
-ARG DIST_PATH
+##
+# base image (abstract)
+##
+FROM python:3.9-slim-bullseye as base
 
+# Build args
+# N.B. The refs updated automagically every release via bumpversion
+ARG dbt_core_ref=dbt-core@v1.0.0
+ARG dbt_postgres_ref=dbt-core@v1.0.0
+ARG dbt_redshift_ref=dbt-redshift@v1.0.0
+ARG dbt_bigquery_ref=dbt-bigquery@v1.0.0
+ARG dbt_snowflake_ref=dbt-snowflake@v1.0.0
+ARG dbt_spark_ref=dbt-spark@v1.0.0
+ARG dbt_spark_version=all
+ARG dbt_third_party
+
+# System setup
 RUN apt-get update \
   && apt-get dist-upgrade -y \
   && apt-get install -y --no-install-recommends \
-  git \
-  ssh-client \
-  software-properties-common \
-  make \
-  build-essential \
-  ca-certificates \
-  libpq-dev \
+    git \
+    ssh-client \
+    software-properties-common \
+    make \
+    build-essential \
+    ca-certificates \
+    libpq-dev \
   && apt-get clean \
-  && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
-
-RUN echo BASE_REQUIREMENTS_SRC_PATH=$BASE_REQUIREMENTS_SRC_PATH
-RUN echo WHEEL_REQUIREMENTS_SRC_PATH=$WHEEL_REQUIREMENTS_SRC_PATH
-RUN echo DIST_PATH=$DIST_PATH
-COPY $BASE_REQUIREMENTS_SRC_PATH ./requirements.txt
-COPY $WHEEL_REQUIREMENTS_SRC_PATH ./wheel_requirements.txt
-COPY $DIST_PATH ./dist
-RUN pip install --upgrade pip setuptools
-RUN pip install --requirement ./requirements.txt
-RUN pip install --requirement ./wheel_requirements.txt
+  && rm -rf \
+    /var/lib/apt/lists/* \
+    /tmp/* \
+    /var/tmp/*
+
+# Env vars
 ENV PYTHONIOENCODING=utf-8
-ENV LANG C.UTF-8
-WORKDIR /usr/app
+ENV LANG=C.UTF-8
+
+# Update python
+RUN python -m pip install --upgrade pip setuptools --no-cache-dir
+
+# Set docker basics
+WORKDIR /usr/app/dbt/
 VOLUME /usr/app
 ENTRYPOINT ["dbt"]
+
+##
+# dbt-core
+##
+FROM base as dbt-core
+RUN python -m pip install "git+https://github.com/dbt-labs/${dbt_core_ref}#egg=dbt-core&subdirectory=core" --no-cache-dir
+
+##
+# dbt-postgres
+##
+FROM base as dbt-postgres
+RUN python -m pip install "git+https://github.com/dbt-labs/${dbt_postgres_ref}#egg=dbt-postgres&subdirectory=plugins/postgres" --no-cache-dir
+RUN python -m pip install "git+https://github.com/dbt-labs/${dbt_core_ref}#egg=dbt-core&subdirectory=core" --no-cache-dir
```

Comment:
> 🌶️ take, I don't think we need to worry about the specifying the dbt-core version for each database adapter specific installation

I completely agree.  That functionality was added because I can see it being useful as a debugging tool, we don't have to build images for every combination (but we could).

Reply:
> I'm unsure how the order of the pip install lines with the --no-cache-dir work in practice as well. The command to install dbt-postgres will use pip to resolve which compatible version of dbt-core to install and all of its dependencies. The command to install dbt-core will install dbt-core with the version using the ref specified and all of its dependencies again. I think combining this on one line will allow pip to install everything and resolve dependency versions once.

So there's sort of two things to unpack here:

1. Do we want to support mix-n-matching db adapter + core versions
  My take:  Yes, if it's possible to do without causing problems for builds that don't mix versions, although I'm happy to re-asses that.
2: Multiple run commands require multiple dependency resolutions, which is inefficient.
  My take:  This is a good point-- I should be able to combine them into a single RUN.

Reply:
1. sounds good to me!
2. I think I more meant that I'm not sure if `pip install dbt-core dbt-postgres` and `pip install dbt-core && dbt-postgres` will result in the same thing. Maybe it is! I haven't dug into pip in some time.

Reply:
Oh I see what you're saying... I'll do some playing around and see if that has the same result.  I still like the idea of putting them in the same RUN layer though.